libdnf/utils/string.cpp

```diff
@@ -0,0 +1,96 @@
+/*
+Copyright (C) 2020-2021 Red Hat, Inc.
+
+This file is part of libdnf: https://github.com/rpm-software-management/libdnf/
+
+Libdnf is free software: you can redistribute it and/or modify
+it under the terms of the GNU Lesser General Public License as published by
+the Free Software Foundation, either version 2.1 of the License, or
+(at your option) any later version.
+
+Libdnf is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU Lesser General Public License for more details.
+
+You should have received a copy of the GNU Lesser General Public License
+along with libdnf.  If not, see <https://www.gnu.org/licenses/>.
+*/
+
+
+#include "string.hpp"
+
+#include <iterator>
+#include <sstream>
+
+
+namespace libdnf::utils::string {
+
+
+std::string join(const std::vector<std::string> & vec, const std::string & delimiter) {
+    std::stringstream result;
+
+    // copy all but the last elements and add delimiter
+    std::copy(vec.begin(), std::prev(vec.end()), std::ostream_iterator<std::string>(result, delimiter.c_str()));
+
+    // if the vector is not empty, add the last element without the delimiter
+    if (!vec.empty()) {
+        result << vec.back();
+    }
+
+    return result.str();
+}
+
+
+std::vector<std::string> rsplit(const std::string & str, const std::string & delimiter, std::size_t max_split) {
+    std::vector<std::string> result;
+
+    if (str.empty()) {
+        result.push_back("");
+        return result;
+    }
+
+    // let's make a copy and resize the string during the split - it's easier than messing with indexes
+    std::string str_copy(str);
+
+    std::size_t current = str_copy.rfind(delimiter);
+    const std::size_t delimiter_length = delimiter.size();
+    while (current != std::string::npos) {
+        if (max_split != std::string::npos && result.size() + 1 > max_split) {
+            result.insert(result.begin(), str_copy);
+            return result;
+        }
+
+        result.insert(result.begin(), str_copy.substr(current + delimiter_length));
+        str_copy.resize(current);
+        current = str_copy.rfind(delimiter);
+    }
+
+    current = 0;
+    result.insert(result.begin(), str_copy);
+    return result;
+}
+
+
+std::vector<std::string> split(const std::string & str, const std::string & delimiter, std::size_t max_split) {
```

Comment:
What is `max_split`? Usually there is a `limit`. If `limit` is set, the returned array will contain a maximum of `limit` elements with the last element containing the whole rest of string. But it seems your split returns array with one extra element. I think, it is inspired by Python. To be honest, I don't like it.